state/factory/factory.go

```diff
@@ -268,7 +266,16 @@ func (sf *factory) newWorkingSet(ctx context.Context, height uint64) (*workingSe
 	if err != nil {
 		return nil, err
 	}
-	store, err := newFactoryWorkingSetStore(sf.protocolView, flusher)
+	var (
+		rootKey    = ArchiveTrieRootKey
+		createTrie = true
+	)
+	if height < sf.currentChainHeight {
+		// archive mode
+		rootKey = fmt.Sprintf("%s-%d", ArchiveTrieRootKey, height)
+		createTrie = false
+	}
+	store, err := newFactoryWorkingSetStore(sf.protocolView, flusher, rootKey, createTrie)
```

Comment:
Splitting into the following two functions seems more clear:
- `newFactoryWorkingSetStore(view, flusher)`, which inner use ArchiveTrieRootKey and true
- `newFactoryWorkingSetStoreAt(view, flusher, height)`, which inner use ArchiveTrieRootKey-height and false

Reply:
does it should use ArchiveTrieRootKey or ArchiveTrieRootKey-height if height equals currentChainHeight?

Reply:
2 cases for height equals currentChainHeight
1. createGenesisStates calls `newWorkingSet(ctx, 0)`, which is write at current tip
2. archive-mode trying to read at current tip height > 0, at the same time `PutBlock` may still be in progress so `ArchiveTrieRootKey-height` may not exist at the moment
both cases should use `ArchiveTrieRootKey `